src/backend/distributed/planner/function_call_delegation.c

```diff
@@ -465,3 +589,202 @@ ShardPlacementForFunctionColocatedWithReferenceTable(CitusTableCacheEntry *cache
 
 	return (ShardPlacement *) linitial(placementList);
 }
+
+
+/*
+ * Checks to see if the procedure is being executed on a worker after delegated
+ * by the coordinator. If the flag forcePushdown is set, capture the distribution
+ * argument value, to be used by the planner to make sure that function uses only
+ * the colocated shards of the distribution argument.
+ */
+void
+CheckDelegatedFunctionExecution(DistObjectCacheEntry *procedure, FuncExpr *funcExpr)
+{
+	/* if the flag is not set, nothing to enforce */
+	if (!procedure->forcePushdown)
+	{
+		return;
+	}
+
+	/*
+	 * On the coordinator PartiallyEvaluateExpression() descends into an
+	 * expression tree to evaluate expressions that can be resolved to a
+	 * constant. Expressions containing a Var are skipped, since the value
+	 * of the Var is not known on the coordinator.
+	 */
+	Node *partitionValueNode = (Node *) list_nth(funcExpr->args,
+												 procedure->distributionArgIndex);
+	Assert(partitionValueNode);
+	partitionValueNode = strip_implicit_coercions(partitionValueNode);
+
+	if (IsA(partitionValueNode, Param))
+	{
+		Param *partitionParam = (Param *) partitionValueNode;
+
+		if (partitionParam->paramkind == PARAM_EXTERN)
+		{
+			/* we should end up here again without a parameter */
+			return;
+		}
+	}
+
+	if (!IsA(partitionValueNode, Const))
+	{
+		/*
+		 * Var node e.g. select fn(col) from table where col=150;
+		 * Param(PARAM_EXEC) node e.g. SELECT fn((SELECT col from test_nested where col=val))
+		 */
+		ereport(ERROR, (errmsg("Non-constant arguments for forcePushdown "
+							   "functions not supported")));
```

Comment:
While exotic, it's not impossible to reach this error:
```
CREATE OR REPLACE FUNCTION test(x int, y bigint)
RETURNS int
AS $$
DECLARE
BEGIN
        RAISE NOTICE '% %', x, y;
        RETURN x + y;
END;
$$ LANGUAGE plpgsql;
SELECT create_distributed_function('test(int,bigint)','$1', force_pushdown := true);

explain analyze with c as (select count(*) from test), b as (select test(random()::int,(select count from c))) select count(*) from b;
ERROR:  Non-constant arguments for forcePushdown functions not supported
CONTEXT:  while executing command on localhost:1400
```

maybe we should just remove it? We can only get here if we are explicitly *not* doing a forced pushdown, so in that case not doing the shard key value checks seems ok.

Reply:
Yeah, removed it (and added that as testcase without explain)